pkg/app_watcher/app_watcher.go

```diff
@@ -0,0 +1,154 @@
+package app_watcher
+
+import (
+	"context"
+	"reflect"
+
+	appclientset "github.com/argoproj/argo-cd/v2/pkg/client/clientset/versioned"
+	"github.com/rs/zerolog/log"
+	"github.com/zapier/kubechecks/pkg/config"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"strings"
+	"time"
+
+	appv1alpha1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	informers "github.com/argoproj/argo-cd/v2/pkg/client/informers/externalversions/application/v1alpha1"
+	applisters "github.com/argoproj/argo-cd/v2/pkg/client/listers/application/v1alpha1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+// ApplicationWatcher is the controller that watches ArgoCD Application resources via the Kubernetes API
+type ApplicationWatcher struct {
+	cfg                  *config.ServerConfig
+	applicationClientset appclientset.Interface
+	appInformer          cache.SharedIndexInformer
+	appLister            applisters.ApplicationLister
+}
+
+// NewApplicationWatcher creates new instance of ApplicationWatcher.
+func NewApplicationWatcher(cfg *config.ServerConfig) (*ApplicationWatcher, error) {
+	// this assumes kubechecks is running inside the cluster
+	kubeCfg, err := clientcmd.BuildConfigFromFlags("", "")
+	if err != nil {
+		log.Fatal().Msgf("Error building kubeconfig: %s", err.Error())
+	}
+
+	appClient := appclientset.NewForConfigOrDie(kubeCfg)
+
+	ctrl := ApplicationWatcher{
+		cfg:                  cfg,
+		applicationClientset: appClient,
+	}
+
+	appInformer, appLister := ctrl.newApplicationInformerAndLister(time.Second * 30)
+
+	ctrl.appInformer = appInformer
+	ctrl.appLister = appLister
+
+	return &ctrl, nil
+}
+
+// Run starts the Application CRD controller.
+func (ctrl *ApplicationWatcher) Run(ctx context.Context, processors int) {
+	log.Info().Msg("starting Application Controller")
+
+	defer runtime.HandleCrash()
+
+	go ctrl.appInformer.Run(ctx.Done())
+
+	if !cache.WaitForCacheSync(ctx.Done(), ctrl.appInformer.HasSynced) {
+		log.Error().Msg("Timed out waiting for caches to sync")
+		return
+	}
+
+	<-ctx.Done()
+}
+
+// onAdd is the function executed when the informer notifies the
+// presence of a new Application in the namespace
+func (ctrl *ApplicationWatcher) onApplicationAdded(obj interface{}) {
+	if !canProcessApp(obj) {
```

Comment:
each time we use this, we end up casting `obj` to `*appv1alpha1.Application`, which this function does too. it'd probably clean things up if you changed `canProcessApp` to return `(*appv1alpha1.Application, bool)`. This would also let you reimplement `cache.MetaNamespaceKeyFunc` to just be `return fmt.Sprintf("%s/%s", app.Metadata.Namespace, app.Metadata.Name)` and avoid the possible error altogether.